### DISCUSSION TITLE
Working with complex custom objectives

### DISCUSSION THREAD

**Discussion by DenisSvirin** [MilesCranmer/PySR#353] [Q&A] [answered]
How can I add and evaluate function to the objective ? 

Essentially, I want to also search for functions for polynomials in your example (_https://astroautomata.com/PySR/examples/#9-custom-objectives_). Something like this `f(P(X)) / g(Q(X))`, where `f = P(X) ** y` or something else. 

I tried to reserve first l and r nodes for these functions and use the rest for polynomials:
```
    tree.degree != 2 && return L(Inf)
   
    attraction_func = tree.l
    repultion_func= tree.r
    
    attraction_func != 2 && return L(Inf)
    repulsion_func != 2 && return L(Inf)
   
    attraction = attraction_func.l
    repulsion = repultion_func.l
```

but it gives `inf` loss and only complexity of 1

**Comment by MilesCranmer** [comment]
Is this
```
attraction_func != 2
```
Supposed to be `attraction_func.degree != 2`?

**Reply by DenisSvirin** [reply]
yes, thanks.
But how i can extract the function?

**Reply by MilesCranmer** [reply]
What do you mean by extract?

**Reply by DenisSvirin** [reply]
As i said i want now to apply functions `attraction_func` and `repulsion_func` to the trees `attraction` and `repulsion`. 

These trees are already just numbers. So i basically have only prediction of these trees and can't use `Node` method.

I suppose i can make new tree with a constant and these functions, but i wonder if there is any easier method

**Reply by MilesCranmer** [reply, marked as answer]
You mean like this?
```julia
    attraction_func = tree.l.r
    attraction = tree.l.l
    attraction_func_val, flag = eval_tree_array(attraction_func, dataset.X, options)
    !flag && return L(Inf)
    attraction_val, flag = eval_tree_array(attraction, dataset.X, options)
    !flag && return L(Inf)

    result = attraction_func_val ^ attraction_val
```
